arch/arm/core/elf.c

```diff
@@ -11,42 +12,265 @@
+	j_two = (*lower >> BIT_THM_BL_J2) & 1;
+	*offset = (sign << SHIFT_THM_JUMPS_SIGN) | ((~(j_one ^ sign) & 1) << SHIFT_THM_BL_J1) |
+				((~(j_two ^ sign) & 1) << SHIFT_THM_BL_J2) |
+				((*upper & MASK_THM_BW_IMM10) << SHIFT_THM_BW_IMM10) |
+				((*lower & MASK_THM_BL_IMM11) << 1);
+	*offset = sign_extend(*offset, SHIFT_THM_JUMPS_SIGN);
+	*offset += sym_base_addr - loc;
 
-	upper &= ARM_BL_BLX_HDR_MASK;
-	lower &= ARM_BL_BLX_HDR_MASK;
-	upper |= (addend >> ARM_BL_BLX_ADDEND_SIZE) & ARM_BL_BLX_ADDEND_MASK;
-	lower |= addend & ARM_BL_BLX_ADDEND_MASK;
+	if (*offset <= THM_JUMP_UPPER_BOUNDARY || *offset >= THM_JUMP_LOWER_BOUNDARY) {
+		LOG_ERR("sym '%s': relocation out of range (%#x -> %#x)\n",
+			sym_name, loc, sym_base_addr);
+		ret = -ENOEXEC;
+	} else {
+		ret = 0;
+	}
 
-	*((uint16_t *)opaddr) = upper;
-	*(((uint16_t *)opaddr) + 1) = lower;
+	return ret;
+}
+
+static inline void thm_jumps_reloc(uint32_t loc, int32_t *offset,
+			uint32_t *upper, uint32_t *lower)
+{
+	uint32_t j_one, j_two, sign;
+
+	sign = (*offset >> SHIFT_THM_JUMPS_SIGN) & 1;
+	j_one = sign ^ (~(*offset >> SHIFT_THM_BL_J1) & 1);
+	j_two = sign ^ (~(*offset >> SHIFT_THM_BL_J2) & 1);
+	*upper = (uint16_t)((*upper & MASK_THM_BW_11110) | (sign << BIT_THM_BW_S) |
+				((*offset >> SHIFT_THM_BW_IMM10) & MASK_THM_BW_IMM10));
+	*lower = (uint16_t)((*lower & (MASK_THM_BL_10|MASK_THM_BL_1)) |
+				(j_one << BIT_THM_BL_J1) | (j_two << BIT_THM_BL_J2) |
+				((*offset >> 1) & MASK_THM_BL_IMM11));
+
+	*(uint16_t *)loc = OPCODE2THM16MEM(*upper);
+	*(uint16_t *)(loc + 2) = OPCODE2THM16MEM(*lower);
+}
+
+static int thm_jumps_handler(elf_word reloc_type, uint32_t loc,
+			uint32_t sym_base_addr, const char *sym_name)
+{
+	int ret;
+	int32_t offset;
+	uint32_t upper, lower;
+
+	ret = thm_jumps_decode(reloc_type, loc, sym_base_addr, sym_name, &offset, &upper, &lower);
+	if (!ret) {
+		thm_jumps_reloc(loc, &offset, &upper, &lower);
+	}
+
+	return ret;
+}
+
+static void thm_movs_handler(elf_word reloc_type, uint32_t loc,
+			uint32_t sym_base_addr, const char *sym_name)
+{
+	int32_t offset;
+	uint32_t upper, lower;
+
+	upper = MEM2THM16OPCODE(*(uint16_t *)loc);
+	lower = MEM2THM16OPCODE(*(uint16_t *)(loc + 2));
+
+	/* MOVT/MOVW instructions encoding in Thumb-2 */
+	offset = ((upper & MASK_THM_MOV_IMM4) << SHIFT_THM_MOV_IMM4) |
+		((upper & MASK_THM_MOV_I) << SHIFT_THM_MOV_I) |
+		((lower & MASK_THM_MOV_IMM3) >> SHIFT_THM_MOV_IMM3) | (lower & MASK_THM_MOV_IMM8);
+	offset = sign_extend(offset, SHIFT_THM_MOVS_SIGN);
+	offset += sym_base_addr;
+
+	if (reloc_type == R_ARM_THM_MOVT_PREL || reloc_type == R_ARM_THM_MOVW_PREL_NC) {
+		offset -= loc;
+	}
+	if (reloc_type == R_ARM_THM_MOVT_ABS || reloc_type == R_ARM_THM_MOVT_PREL) {
+		offset >>= 16;
+	}
+
+	upper = (uint16_t)((upper & (MASK_THM_MOV_11110|MASK_THM_MOV_100100)) |
+		((offset & (MASK_THM_MOV_IMM4<<SHIFT_THM_MOV_IMM4)) >> SHIFT_THM_MOV_IMM4) |
+		((offset & (MASK_THM_MOV_I<<SHIFT_THM_MOV_I)) >> SHIFT_THM_MOV_I));
```

Comment:
I've tried to address all your requests but I don't get this one. I checked the FIELD_PREP()/FIELD_GET() and I got scared :) They contains division/multiplication and, in my case, they would be used with a variable so... I don't know how the compiler could optimize it... Do we really need to use these macros here?!

Reply:
Sorry for the delay, missed this reply last week! Not a deal breaker, it was always in the same spirit of using "native" Zephyr tools. These macros use a maths trick (multiplication / division by a power of 2  - the `LSB_GET` returns the lowest bit _in the bitmask_) to actually shift numbers to/from offset 0; the compiler turns this into a constant shift operation, so it has no runtime cost.
The macros also automatically chop bits outside the mask you give them, so you can avoid some typing.

The current code uses the following numbers:
```C
#define MASK_THM_MOV_I BIT(10) 
#define MASK_THM_MOV_IMM4 GENMASK(3, 0) 
#define SHIFT_THM_MOV_I 1 
#define SHIFT_THM_MOV_IMM4 12 
...
((offset & (MASK_THM_MOV_IMM4<<SHIFT_THM_MOV_IMM4)) >> SHIFT_THM_MOV_IMM4) |
((offset & (MASK_THM_MOV_I<<SHIFT_THM_MOV_I)) >> SHIFT_THM_MOV_I)
```
I would rewrite it as: 
```C
#define MASK_THM_MOV_I BIT(10) 
#define MASK_THM_MOV_IMM4 GENMASK(3, 0) 
#define SHIFT_THM_MOV_I 11 // <--
#define SHIFT_THM_MOV_IMM4 12
... 
FIELD_PREP(MASK_THM_MOV_IMM4, offset >> SHIFT_THM_MOV_IMM4) | 
FIELD_PREP(MASK_THM_MOV_I, offset >> SHIFT_THM_MOV_I)
```
Note how the shift for `MOV_I` is 11, which I find easier to check and maintain than "the position of the mask, plus or minus 1 depending on the direction" (since some have inverted shifts!). 
(the one for `MOV_IMM4` gets lucky as the mask is zero-based, so the shift does not change :slightly_smiling_face:)